pipeline/trusted-resources/pkg/trustedtask/sign_test.go

```diff
@@ -0,0 +1,214 @@
+/*
+Copyright 2022 The Tekton Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package trustedtask
+
+import (
+	"bytes"
+	"context"
+	"encoding/base64"
+	"net/http/httptest"
+	"net/url"
+	"testing"
+
+	"github.com/google/go-containerregistry/pkg/authn/k8schain"
+	"github.com/google/go-containerregistry/pkg/registry"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+	"github.com/sigstore/sigstore/pkg/signature"
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
+)
+
+func TestSignTaskSpec(t *testing.T) {
+	ctx := context.Background()
+	sv, err := GetSignerVerifier(password)
+	if err != nil {
+		t.Fatalf("failed to get signerverifier %v", err)
+	}
+
+	tcs := []struct {
+		name     string
+		signer   signature.SignerVerifier
+		taskSpec *v1beta1.TaskSpec
+		wantErr  bool
+	}{{
+		name:     "Sign TaskSpec",
+		signer:   sv,
+		taskSpec: taskSpecTest,
+		wantErr:  false,
+	}, {
+		name:     "Empty TaskSpec",
+		signer:   sv,
+		taskSpec: nil,
+		wantErr:  false,
+	}, {
+		name:     "Empty Signer",
+		signer:   nil,
+		taskSpec: taskSpecTest,
+		wantErr:  true,
+	},
+	}
+
+	for _, tc := range tcs {
+		t.Run(tc.name, func(t *testing.T) {
+			sig, err := SignTaskSpec(tc.signer, taskSpecTest)
+			if (err != nil) != tc.wantErr {
+				t.Fatalf("SignTaskSpec() get err %v, wantErr %t", err, tc.wantErr)
+			}
+
+			if tc.wantErr {
+				return
+			}
+			signature, err := base64.StdEncoding.DecodeString(sig)
+			if err != nil {
+				t.Fatalf("error decoding signature: %v", err)
+			}
+			if err := VerifyTaskSpec(ctx, taskSpecTest, tc.signer, signature); err != nil {
```

Comment:
New thread from https://github.com/tektoncd/experimental/pull/846#discussion_r826329155 because GitHub won't let me respond to a thread on an orphaned commit 😭 

I'm not concerned about the input types. By testing Sign with Verify, we're ensuring that they can be used with each other without error (which is good to test!), but it doesn't tell us whether they are necessarily doing the right things by themselves. It creates some soft spots in testing because of the circular dependency - i.e. we can only trust that the Verify funcs are correct if the Sign funcs are correct, but we can only trust that the Sign funcs are correct if the Verify funcs are correct.

What would be good to add is a baseline to build trust that at least one of them can be verified independently of the other (similar to an inductive proof), even if it means introducing 1-2 brittle tests (i.e. compare against a precomputed signature). My guess is Sign is probably the better target here, because then it's safe to reuse in verify.

Reply:
Ah yes.
So if I directly call the `signer.SignMessage()`,  and `verifier.VerifySignature()` from cosign, then I don't have to do the tests. Because they are third-party functions and should be tested there? 

But I wrap some customized logic to create new `Sign()` and `Verify()`, and as you mentioned they have the same internal logic and depend on each other for testing.

Reply:
What if when testing `SignTaskSpec()` I don't directly call `VerifyTaskSpec` but instead call `verifier.VerifySignature`? Will that break the  circular dependency?
I'm not sure if use precomputed signature can work, since even with the same key, every time we generate the signature, the signature is different.

Reply:
That would be okay - but we should be careful to avoid duplicating logic that's in the Verify funcs. If we end up reimplementing what's done in VerifyTaskSpec in the test, that's not much better than just calling it - any complexity added to the Verify functions will need to be added to the test.

I think the easiest thing to do would be to test Sign in isolation with a select few outputs that we've manually verified are correct. This should give us enough confidence to let us use the function elsewhere.

Reply:
So I added a mock signer, which just return the content, not sure if this makes sense